src/ipnet.rs

```diff
@@ -0,0 +1,828 @@
+    pub fn new(addr: Ipv4Addr, width: u8) -> Result<Self, IpNetPrefixError> {
+        if width > IPV4_NET_WIDTH_MAX {
+            Err(IpNetPrefixError(width))
+        } else {
+            Ok(Self { addr, width })
+        }
+    }
+
+    /// Create an Ipv4Net that contains *exclusively* the given address.
+    pub fn host_net(addr: Ipv4Addr) -> Self {
+        Self {
+            addr,
+            width: IPV4_NET_WIDTH_MAX,
+        }
+    }
+
+    /// Return the base address used to create this subnet.
+    pub fn addr(&self) -> Ipv4Addr {
+        self.addr
+    }
+
+    /// Return the prefix width.
+    pub fn width(&self) -> u8 {
+        self.width
+    }
+
+    pub(crate) fn mask(&self) -> u32 {
+        u32::MAX << (IPV4_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return true iff the subnet contains only the base address i.e. the
+    /// size is exactly one address.
+    pub fn is_host_net(&self) -> bool {
+        self.width == IPV4_NET_WIDTH_MAX
+    }
+
+    /// Return the number of addresses contained within this subnet.
+    pub fn size(&self) -> u32 {
+        1u32 << (IPV4_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return the prefix address (the base address with the mask applied).
+    pub fn prefix(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr & self.mask())
+    }
+
+    /// Return the network address for subnets wider than a /31 or /32.
+    pub fn network(&self) -> Option<Ipv4Addr> {
+        (self.width < 31).then(|| {
+            let addr: u32 = self.addr.into();
+            Ipv4Addr::from(addr & self.mask())
+        })
+    }
+    /// Return the broadcast address for subnets wider than a /31 or /32.
+    pub fn broadcast(&self) -> Option<Ipv4Addr> {
+        (self.width < 31).then(|| {
+            let addr: u32 = self.addr.into();
+            Ipv4Addr::from(addr | !self.mask())
+        })
+    }
+
+    /// Return the first address within this subnet.
+    pub fn first_addr(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr & self.mask())
+    }
+
+    /// Return the last address within this subnet.
+    pub fn last_addr(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr | !self.mask())
+    }
+
+    /// Return the first host address within this subnet; for subnets wider
+    /// than a /31 or /32 that is the first address after the network address.
+    pub fn first_host(&self) -> Ipv4Addr {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let first = addr & mask;
+        if self.width == 31 || self.width == 32 {
+            Ipv4Addr::from(first)
+        } else {
+            Ipv4Addr::from(first + 1)
+        }
+    }
+
+    /// Return the last host address within this subnet; for subnets wider
+    /// than a /31 or /32 that is the address immediately before the broadcast
+    /// address.
```

Comment:
I think we should state what this returns if the width is >= `31`.

Reply:
reworked.